### DISCUSSION TITLE
achieving auto-complete - partial success, usability and features

### DISCUSSION THREAD

**Discussion by shula** [beeware/toga#2705] [General]
I'm using autocomplete in many forms/apps. It saves browsing time, typos, and extra clicks.
Let's explore the options;

1) Toga hasn't got a built-in way to achieve autocomplete AFAIK.
2) On windows it is often done withing a ComboBox.
3) Toga doesn't have it comboboxes. 
4) The closest is "selection" box, but it's not dropping down, and is not editable. It's not even a ListBox..

## What I did (see screenshot)

A TextInput on top,
a DetailedList directly below.
TextInput.on_change()  --> delete existing list items, creates a filtered list, according to the text.
TextInput.on_confirm()  --> to capture the first item on the list.

It works. 
but there are UI/UX issues I want to fix:

1. There's no way to "confirm" a selected list-item.   I want to do something when the user to **press ENTER** and/or **double click** on a list item. 
2. there's a "on_primary_action()" - but it does does not respond to ENTER, or double-click.
3. I can't assign "Down Arrow" to the textInput, to change focus.  Tab key does that, but regular users don't know "tab" .
4. I suspect this re-writing the items has time penalty, when the data grows. 
5. the auto-complete options widget is constantly visible. I want to open it only when it's relevant.
 
![autocomplete-example](https://github.com/beeware/toga/assets/124651/0d2594e6-091e-4689-9d80-ae120072ca4b)

Is there a solution to either points?

**Comment by mhsmith** [comment]
> There's no way to "confirm" a selected list-item.   I want to do something when the user to **press ENTER** and/or **double click** on a list item.

DetailedList currently looks like a table on Windows, but that may change in the future. So if you want it to look like a table, you should use the [Table](https://toga.readthedocs.io/en/stable/reference/api/widgets/table.html) widget instead.

Table already has an `on_activate` event which responds to a double-click, but it doesn't respond to Enter. I've created an issue for that: #2707.

Meanwhile, you can already respond to a single click with the `on_select` event.

> the auto-complete options widget is constantly visible. I want to open it only when it's relevant

Widgets can be added to and removed from the window at any time using the methods of the [WIdget base class](https://toga.readthedocs.io/en/stable/reference/api/widgets/widget.html).

**Comment by shula** [comment]
thank you.
That will be helpful.

(on a side note:
10 On_select doesn't solve it while using the keyboard - because every movement triggers a "select" action,
and a tabe doesn't seem like a solution for a combobox-dropdown, but maybe with some careful styling.)
